File: src/main/scala/net/zenxarch/bot/defense/DefenseStateManager.scala
package net.zenxarch.bot.defense

import net.minecraft.client.gui.screen.ingame.HandledScreen
import net.zenxarch.bot.defense.modules._
import net.zenxarch.bot.util.TargetUtil
import net.zenxarch.bot.ZenBot.mc

object DefenseStateManager {
  private var isDefenseActive = false
  private var isActionPerformed = false
  private var modules: List[Module] = null

  private var shouldCloseHandledScreen = false

  def setActiveStatus(status: Boolean) = {
    isDefenseActive = status
  }

  def getActiveStatus() = isDefenseActive

  def init() = {
    TargetUtil.init()
    modules = List(
      new WaterMLG(),
      new KillAura(),
      new AutoFire(),
      new ShieldBlock()
    )
  }

  private def checkTargets() = TargetUtil.getNearestEnemyPlayer() == null &&
    TargetUtil.getNearestHostile() == null &&
    TargetUtil.getNearestPassive() == null &&
    TargetUtil.getNearestProjectile() == null

  private def allInactive() = modules.filter(_.isActive()).length == 0

  def preTick(): Unit = {
    if (!isDefenseActive)
      return
    isActionPerformed = false

    if (allInactive())
      return

    TargetUtil.updateTargets()
    if (checkTargets() && !tryHandleMcScreen())
      return

    forEachModule(_.preTick())

    var projectileTarget = TargetUtil.getNearestProjectile()
    if (projectileTarget != null) {
      forEachModule(_.handleProjectile(projectileTarget))
      return
    }

    var hostileTarget = TargetUtil.getNearestHostile()
    if (hostileTarget != null) {
      forEachModule(_.handleHostile(hostileTarget))
      return
    }

    var playerTarget = TargetUtil.getNearestEnemyPlayer()
    if (playerTarget != null) {
      forEachModule(_.handlePlayer(playerTarget))
      return
    }

    var passiveTarget = TargetUtil.getNearestPassive()
    if (passiveTarget != null) {
      forEachModule(_.handlePassive(passiveTarget))
      return
    }

    forEachModule(_.handleNone())
  }

  private def forEachModule(exec: (Module) => Unit) = {
    for
      module <- modules
      if module.isActive()
    do exec(module)
  }

  def postTickCheck() = isActionPerformed

  def performAction(action: () => Boolean): Boolean = {
    if (isActionPerformed)
      return false
    isActionPerformed = action.apply()
    return isActionPerformed
  }

  def tryHandleMcScreen(): Boolean = {
    if (mc.currentScreen == null)
      return true
    if (mc.currentScreen.isInstanceOf[HandledScreen]) {
      if (shouldCloseHandledScreen)
        mc.player.closeHandledScreen()
      return shouldCloseHandledScreen
    }
    return true
  }
}

File: src/main/java/net/zenxarch/bot/defense/DefenseStateManager.java
package net.zenxarch.bot.defense;

import static net.zenxarch.bot.ZenBot.mc;

import java.util.ArrayList;
import java.util.function.BooleanSupplier;
import java.util.function.Consumer;
import net.minecraft.client.gui.screen.ingame.HandledScreen;
import net.zenxarch.bot.defense.modules.*;
import net.zenxarch.bot.defense.modules.Module;
import net.zenxarch.bot.util.TargetUtil;

public class DefenseStateManager {
  private static boolean isDefenseActive;
  private static boolean isActionPerformed;
  private static final ArrayList<Module> modules = new ArrayList<>();

  private static boolean shouldCloseHandledScreen;

  public static void setActiveStatus(boolean status) {
    isDefenseActive = status;
  }

  public static boolean getActiveStatus() { return isDefenseActive; }

  public static void init() {
    TargetUtil.init();
    modules.add(new WaterMLG());
    modules.add(new KillAura());
    modules.add(new AutoFire());
    modules.add(new ShieldBlock());
  }

  private static boolean checkTargets() {
    return TargetUtil.getNearestEnemyPlayer() == null &&
        TargetUtil.getNearestHostile() == null &&
        TargetUtil.getNearestPassive() == null &&
        TargetUtil.getNearestProjectile() == null;
  }

  private static boolean allInactive() {
    for (var module : modules) {
      if (module.isActive())
        return false;
    }
    return true;
  }

  public static void preTick() {
    if (!isDefenseActive)
      return;
    isActionPerformed = false;

    if (allInactive())
      return;

    TargetUtil.updateTargets();
    if (checkTargets() && !tryHandleMcScreen())
      return;

    forEachModule(m -> m.preTick());

    var projectileTarget = TargetUtil.getNearestProjectile();
    if (projectileTarget != null) {
      forEachModule(m -> m.handleProjectile(projectileTarget));
      return;
    }

    var hostileTarget = TargetUtil.getNearestHostile();
    if (hostileTarget != null) {
      forEachModule(m -> m.handleHostile(hostileTarget));
      return;
    }

    var playerTarget = TargetUtil.getNearestEnemyPlayer();
    if (playerTarget != null) {
      forEachModule(m -> m.handlePlayer(playerTarget));
      return;
    }

    var passiveTarget = TargetUtil.getNearestPassive();
    if (passiveTarget != null) {
      forEachModule(m -> m.handlePassive(passiveTarget));
      return;
    }

    forEachModule(m -> m.handleNone());
  }

  private static void forEachModule(Consumer<? super Module> exec) {
    for (var module : modules) {
      if (module.isActive()) {
        exec.accept(module);
      }
    }
  }

  public static boolean postTickCheck() { return isActionPerformed; }

  public static boolean performAction(BooleanSupplier action) {
    if (isActionPerformed)
      return false;
    isActionPerformed = action.getAsBoolean();
    return isActionPerformed;
  }

  public static boolean tryHandleMcScreen() {
    if (mc.currentScreen == null)
      return true;
    if (mc.currentScreen instanceof HandledScreen) {
      if (shouldCloseHandledScreen)
        mc.player.closeHandledScreen();
      return shouldCloseHandledScreen;
    }
    return true;
  }
}
